pyrtls.pyi

```diff
@@ -0,0 +1,407 @@
+are pretty new and not yet feature-complete.
+"""
+
+from collections.abc import Iterable
+from socket import socket
+
+# lib.rs
+
+class TrustAnchor:
+    def __new__(
+        cls,
+        subject: bytes,
+        subject_public_key_info: bytes,
+        name_constraints: bytes | None,
+    ) -> TrustAnchor: ...
+
+class IoState:
+    def tls_bytes_to_write(self) -> int:
+        """
+        How many bytes could be written by `Connection.write_tls_into()` if called right now.
+        A non-zero value implies that `Connection.wants_write()` would yield `True`.
+        """
+
+    def plaintext_bytes_to_read(self) -> int:
+        """How many plaintext bytes are currently buffered in the connection."""
+
+    def peer_has_closed(self) -> bool:
+        """
+        `True` if the peer has sent us a `close_notify` alert. This is the TLS mechanism to
+        securely half-close a TLS connection, and signifies that the peer will not send any
+        further data on this connection.
+        """
+
+class TLSError: ...
+
+# client.rs
+
+class ClientSocket:
+    """
+    A `ClientSocket` is a wrapper type that contains both a `socket.socket` and a
+    `ClientConnection` object. It is similar to the `ssl.SSLSocket` class from the
+    standard library and should implement most of the same methods.
+    """
+
+    def connect(self, address: tuple[str, int]) -> None:
+        """
+        Connect to a remote socket address. `address` must currently be a 2-element
+        tuple containing a hostname and a port number.
+        """
+
+    def do_handshake(self) -> None:
+        """Perform the TLS setup handshake."""
+
+    def send(self, bytes: bytes) -> int:
+        """
+        Send data to the socket. The socket must be connected to a remote socket. Returns the
+        number of bytes sent. Applications are responsible for checking that all data has been
+        sent; if only some of the data was transmitted, the application needs to attempt delivery
+        of the remaining data.
+        """
+
+    def recv(self, size: int) -> bytes:
+        """
+        Receive data from the socket. The return value is a bytes object representing the data
+        received. The maximum amount of data to be received at once is specified by `size`.
+        A returned empty bytes object indicates that the server has disconnected.
+        """
+
+class ClientConnection:
+    """
+    A `ClientConnection` contains TLS state associated with a single client-side connection.
+    It does not contain any networking state, and is not directly associated with a socket,
+    so I/O happens via the methods on this object directly.
+
+    A `ClientConnection` can be created from a `ClientConfig` `config` and a server name, `name`.
+    The server name must be either a DNS hostname or an IP address (only string forms are
+    currently accepted).
+    """
+
+    def __new__(cls, config: ClientConfig, name: str) -> ClientConnection: ...
+    def readable(self) -> bool:
```

Comment:
Nit: probably good to add an empty line between these methods?

Reply:
My code formatter (black) removes it, but it's your choice.